articles/ros2_threat_model.md

```diff
@@ -0,0 +1,3291 @@
+    <td class="danger">✘</td>
+    <td class="success">✓</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">3</td>
+    <td class="danger">3</td>
+    <td class="warning">2</td>
+    <td class="danger">3</td>
+    <td class="danger">3</td>
+    <td>14</td>
+    <td class="danger">✘</td>
+    <td class="success">✓</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="success">✓</td>
+    <td class="success">✓</td>
+    <td class="warning">▲</td>
+    <td class="success">✓</td>
+    <td class="success">✓</td>
+    <td class="warning">▲</td>
+    <td class="warning">▲</td>
+    <td>
+      <ul>
+        <li>Deployment service should be authenticated• Communication with
+the deployment service should be done over a secure channel.</li>
+      </ul>
+    </td>
+    <td> </td>
+  </tr>
+
+  <tr>
+    <td>An attacker intercepts the binaries sent by the depoyment service.</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="success">✓</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="success">1</td>
+    <td class="danger">3</td>
+    <td class="warning">2</td>
+    <td class="danger">3</td>
+    <td class="danger">3</td>
+    <td>12</td>
+    <td class="danger">✘</td>
+    <td class="success">✓</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="success">✓</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="success">✓</td>
+    <td>
+      <ul>
+        <li>Deployment service should be authenticated• Communication with
+the deployment service should be done over a secure channel.</li>
+      </ul>
+    </td>
+    <td> </td>
+  </tr>
+
+  <tr>
+    <td>An attacker prevents the robot and the deployment service from
+communicating.</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="success">✓</td>
+    <td class="danger">✘</td>
+    <td class="success">1</td>
+    <td class="danger">3</td>
+    <td class="warning">2</td>
+    <td class="danger">3</td>
+    <td class="danger">3</td>
+    <td>12</td>
+    <td class="danger">✘</td>
+    <td class="success">✓</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="success">✓</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    <td class="danger">✘</td>
+    
```

Comment:
How about adding another section to treat vulnerabilities in the _Communications Middleware_? Something like "Embedded / Software / Communication Middleware". There're several issues opened in ros2/ros2 describing how middleware implementations start demanding a considerable amount of resources as more publishers (or subscribers) are enabled into them. We ourselves while building robots found considerable limitations in several DDS implementations. This can lead to simple DoS attacks by simply accessing the local network of the robot. While we haven't yet tested it extensively (@LanderU, feel free to comment), we've been able to limit the implications to a certain extend by properly configuring the DDS layer.

Reply:
I think this would fall on "An attacker prevents a communication channel from being usable."

Reply:
@vmayoral My understanding is that enabling SROS and enforcing all nodes to be secure would mitigate this. Maybe adding somewhere "the DDS network should be kept isolated as much as possible" would be good?

By the way, we've been running TSan on ROS 2 and we detected potential deadlocks when the graph changes. So what you're highlighting here is a real concern, due to the way DDS discovery works, one could imagine that spamming a network with new DDS participants could put ROS 2 nodes into a dead-locked state (note that we didn't attempt this yet so we don't know how hard it would be to reproduce this issue that TSan detected).